Challenge/score

LukaszTheLittle opened the pull request:
This feature shows toast at the end of quiz with score for correct answers

LukaszTheLittle committed:
Added new variable to data class `Question`

LukaszTheLittle committed:
Added variable `isCurrentQuestionAnswered`
Added function `questionIsAnswered`

LukaszTheLittle committed:
Implemented function `isAnswered` that disables buttons `true / false`

LukaszTheLittle committed:
indentation correction

LukaszTheLittle committed:
deleted redundant space

LukaszTheLittle committed:
Renamed method `isAnswered` to `updateButtonsEnabledState` to be more precise

LukaszTheLittle committed:
Added variable `questionBankSize`
Added variable `score`

LukaszTheLittle committed:
Created function `startCheatActivity()`
Created function `pointForCorrectAnswer()` and `if statement` in that adds point for every correct answer

LukaszTheLittle committed:
Added variable `questionBankAreAnswered`

LukaszTheLittle committed:
Added `scoreToast` which shows how much points user got for correct answers
Added `if statement` which shows `toastAnswerStatus` or `scoreToast`

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/geoquiz/MainActivity.kt:
```diff
@@ -121,22 +122,68 @@ class MainActivity : AppCompatActivity() {
+            pointForCorrectAnswer()
+        }
+
+        val toastAnswerStatus = Toast.makeText(
+            this,
+            messageResId,
+            Toast.LENGTH_SHORT
+        )
+        toastAnswerStatus.setGravity(Gravity.TOP, 0, 300)
+        toastAnswerStatus.show()
+
+        val scoreToast = Toast.makeText(
+                this,
+                "Twój wynik to " +
+                        "${(quizViewModel.score / quizViewModel.questionBankSize * 100).toInt()}%",
+                Toast.LENGTH_LONG
+            )
+        scoreToast.setGravity(Gravity.TOP, 0, 150)
+
+        if (quizViewModel.questionBankAreAnswered) {
+            toastAnswerStatus.show()
+            scoreToast.show()
+        } else {
+            toastAnswerStatus.show()
+        }
```
it would be better to extract this to a separate method, i.e `showScoreToast`. You want to avoid methods containing tens of lines of code, better to keep things simple and separated. It's also easier to test after - imagine if you wanted to test only the code that shows a toast. With the code above, you would have to call `checkAnswer` each time to test the toast, and have potentially unwanted results.

LukaszTheLittle replied:
done!

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/geoquiz/MainActivity.kt:
```diff
@@ -121,22 +122,68 @@ class MainActivity : AppCompatActivity() {
+    }
+
+    private fun pointForCorrectAnswer() {
+        quizViewModel.score++
```
normally, you would do this differently. You have learned about encapsulation already. If you have a component that exposes some data, its better to mutate this data only from within. On this example, that would mean to have a `private` method in the viewModel, i.e `incrementScore` and a public method `updateScore`. Then from here you would call `updateScore` only, but in the viewModel you would have:
```
fun updateScore() {
   this.incrementScore()
}

private fun incrementScore() {
   this.score++
}
```

This is important, as it defines the contract between components, meaning what you can and can't do. Having a public `updateScore` you do not directly mutate any state of it, but delegate it to it's internal functions. It also helps with testing later on

LukaszTheLittle replied:
Thanks, made changes as you suggested

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/geoquiz/Question.kt:
```diff
@@ -2,4 +2,7 @@ package com.bignerdranch.android.geoquiz
 
-data class Question(@StringRes val textResId: Int, val answer: Boolean)
\ No newline at end of file
+data class Question(@StringRes
+                    val textResId: Int,
+                    val answer: Boolean,
+                    var isAnswered: Boolean = false)
```
wrong indentation

LukaszTheLittle replied:
corrected

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/geoquiz/QuizViewModel.kt:
```diff
@@ -5,6 +5,7 @@ import androidx.lifecycle.ViewModel
 class QuizViewModel : ViewModel() {
 
     var currentIndex = 0
+    var score = 0.0
```
this goes to this variable and the others in this class:

if you are exposing them to outside, its better to expose a final `val` that cannot be mutated, which internally (from here) references a mutable variable, i.e:
```
private var scoreData = 0.0
val score get() = scoreData
```
that way you are ensuring that you can _only_ read that property from the outside, but do not have the direct access to its state. This comes in handy down the line, when your code gains on complexity

LukaszTheLittle replied:
done

Pantkowsky commented on app/src/main/java/com/bignerdranch/android/geoquiz/QuizViewModel.kt:
```diff
@@ -21,6 +22,15 @@ class QuizViewModel : ViewModel() {
     val currentQuestionText: Int
         get() = questionBank[currentIndex].textResId
 
+    val isCurrentQuestionAnswered: Boolean
+        get() = questionBank[currentIndex].isAnswered
+
+    val questionBankSize: Int
+        get() = questionBank.size
+
+    val questionBankAreAnswered: Boolean
+        get() = questionBank.all { it.isAnswered }
```
organize your `val`s and `private val`s here. You want to keep them grouped, i.e all `private`s together, and all publics together. It's easier to then keep track of what is what

LukaszTheLittle replied:
Organized them

LukaszTheLittle committed:
Indentation correction

LukaszTheLittle committed:
Removed single lambda parameter declarations

LukaszTheLittle committed:
Organized `private val`s, `val`s, `var`s and methods

LukaszTheLittle committed:
Changed `var score` to `private`
Added `private fun incrementScore` method
Added public method `updateScore`
Made appropriate changes in `MainActivity.kt`'s `pointForCorrectAnswer` method

LukaszTheLittle committed:
Little change to `val getScore`

LukaszTheLittle committed:
`supportsRtl` changes

LukaszTheLittle committed:
Added `arrow_left.xml` and `arrow_right.xml` instead of png arrow pictures

LukaszTheLittle committed:
Added newer version of `'com.google.android.material:material:1.6.0'` implementation

LukaszTheLittle committed:
Changed `Buttons` to `androidx.appcompat.widget.AppCompatButton`

LukaszTheLittle committed:
Added `button_background.xml` for button's background

LukaszTheLittle committed:
Created data class `AnswerData` and val `answerData`

LukaszTheLittle committed:
Updated `checkAnswer` function parameters in `trueButton` and `falseButton`
Reduced `checkAnswer` function to be more legible
Created `answerStatusToast` method
Created `scoreToast` method
Created `showToast` method
Created `const val JUDGEMENT_TOAST`
Created `const val CORRECT_TOAST`
Created `const val INCORRECT_TOAST`

LukaszTheLittle committed:
Updated `activity_main` for landscape mode

LukaszTheLittle committed:
Corrected `layout_constraintWidth_percent` for `next_button` and `prev_button`

LukaszTheLittle committed:
Added `layout_marginTop` to `TextView`

LukaszTheLittle committed:
Corrected indentation

LukaszTheLittle committed:
`var currentIndex` changed to be private
Created `val getCurrentIndex`
Deleted unused `val` from data class `AnswerData`
Deleted unused `isCheaterStatus` property

LukaszTheLittle committed:
Deleted unused property `currentIndex`
Deleted unused parameter in `showToast()` method
Corrected if statement in `showToast()` method
Modified parameter value in `savedInstanceState.putInt()`

LukaszTheLittle commented:
This PR was instructive! Had a lot of fun :)
